src/app/CommandHandler.cpp

```diff
@@ -536,10 +609,14 @@ CHIP_ERROR CommandHandler::FinishCommand(bool aStartDataStruct)
     {
         ReturnErrorOnFailure(commandData.GetWriter()->EndContainer(mDataElementContainerType));
     }
+
+    if (mRefForResponse.HasValue())
+    {
+        ReturnErrorOnFailure(commandData.Ref(mRefForResponse.Value()));
+    }
+
     ReturnErrorOnFailure(commandData.EndOfCommandDataIB());
     ReturnErrorOnFailure(mInvokeResponseBuilder.GetInvokeResponses().GetInvokeResponse().EndOfInvokeResponseIB());
-    ReturnErrorOnFailure(mInvokeResponseBuilder.GetInvokeResponses().EndOfInvokeResponses());
```

Comment:
It is arguable if this is the right approach here. I am looking for feedback. We might want to always end the InvokeResponses and InvokeRsponseMessages, and simply "rollback" if we are going to add something else.

For the time being I did not do that as it adds some complexity, and just went with keep it simple

Reply:
I will mark this comment as resolved and any further discussion can be in https://github.com/project-chip/connectedhomeip/pull/30614#discussion_r1403672212